java/code/src/com/suse/manager/webui/controllers/ProductsController.java

```diff
@@ -237,25 +240,33 @@ public static String addProduct(Request request, Response response, User user) {
         ContentSyncManager csm = new ContentSyncManager();
         if (csm.isRefreshNeeded(null)) {
             log.fatal("addProduct failed: Product Data refresh needed");
-            json(response, identifiers.stream().collect(Collectors.toMap(ident -> ident, ident -> Boolean.FALSE)));
+            return json(response, identifiers.stream().collect(Collectors.toMap(Function.identity(), ident -> Map.of(
+                "success", false,
+                "error", LocalizationService.getInstance().getMessage("setup.product.error.dataneedsrefresh")
+            ))));
         }
-        if (log.isDebugEnabled()) {
-            log.debug("Add/Sync products: " + identifiers);
-        }
-        Map<String, Optional<? extends Throwable>> stringOptionalMap =
-                new ProductSyncManager().addProducts(identifiers, user);
-
-        stringOptionalMap.forEach((k, v) -> {
-            v.ifPresent(error -> {
-                log.fatal("addProduct failed for " + k, error);
-            });
-        });
 
-        Map<String, Boolean> collect = stringOptionalMap.entrySet().stream().collect(Collectors.toMap(
-                Map.Entry::getKey,
-                kv -> kv.getValue().isPresent()
-        ));
-        return json(response, collect);
+        LogMF.debug(log, "Add/Sync products: {}", identifiers);
+
+        ProductSyncManager psm = new ProductSyncManager();
+        Map<String, Optional<? extends Exception>> productStatusMap = psm.addProducts(identifiers, user);
+
+        // Convert to a map specifying operation result for each product while logging the errors that have happened
+        Map<String, HashMap<Object, Object>> resultMap = productStatusMap.entrySet().stream().collect(
+            Collectors.toMap(
+            Map.Entry::getKey, entry -> {
+                // Return the outcome and the error message to give a better notification in the frontend
+                final HashMap<Object, Object> productResult = new HashMap<>();
```

Comment:
The key here seems to be always a String so we can reflect that here.

Reply:
also thinking about it the presence or absence of error should be enough to determine if it was successful so unless this format is already something we were using before i would get rid of the success flag.

Reply:
Before we were using just the boolean value in the frontend. I could indeed just return the error message and have it as `null` if there was no error. I'm not sure though if returning null for a successful is clean, from a JSON response point of view. It is anyway totally doable, I just have to change a bit more the typescript code.

Reply:
I see json here just as the wire format and from that perspective i personally don't care how it looks like as long as its efficient and leaves little room for error. In this case i'm just mentioning it because the 2 fields carry the same information and you can create a response where they say different things which makes little sense. I think we should ask the frontend people what they think about since they have to handle it.

Reply:
@Etheryte What do you think?

Reply:
Jumping in with fairly little context here, so take my opinion with a grain of salt. Imo the current boolean value approach is not as good as it could be, as understanding it requires either the comments to be up-to-date or for the developer to jump between the frontend and backend code. From that perspective, either one of having a success flag or an error message field would work well to avoid that problem, I don't have a strong preference for either. I don't see any extra merit to having both, since you would probably end up using only one to check anyway.

Reply:
I refactored the code and get rid of the flag returning only the error message, if present. The downside of this is that with GSON we are actually not serializing null values, so in the end the response contains only the failed products. This does not affect much the logic on the frontend side and the code is able to handle both null and missing values.